addons/l10n_id/models/account_move.py

```diff
@@ -0,0 +1,152 @@
+
+    def _cron_update_payment_status(self):
+        """
+        This cron will:
+            - Get all invoices that are not paid, and have details about QRIS qr codes.
+            - For each invoices, get information about the payment state of the QR using the API.
+            - If the QR is not paid and it has been more than 30m, we discard that qr id (no longer valid)
+            - If it is paid, we will register the payment on the invoices.
+        """
+        invoices = self.search([
+            ('payment_state', '=', 'not_paid'),
+            ('l10n_id_qris_invoice_details', '!=', False)
+        ])
+        return invoices._update_payment_status()
+
+    def action_update_payment_status(self):
+        """ Do the same thing as the cron, but triggered manually. If payments are created, they will be displayed. """
+        invoices = self.filtered_domain([
+            ('payment_state', '=', 'not_paid'),
+            ('l10n_id_qris_invoice_details', '!=', False)
+        ])
+        return invoices._update_payment_status()
+
+    def _update_payment_status(self):
+        qr_statuses = self._get_qris_qr_statuses()
+
+        # For unpaid invoices, we will just discard every QRIS QR code that is expired, and then fetch the status of the
+        # remaining ones next time.
+        self._process_unpaid_invoices({invoice: status for invoice, status in qr_statuses.items() if status['paid'] is False})
+
+        # For the paid invoices, we will discard the stored QRIS data, log a message and register the payment.
+        # Return the newly created payments, if any.
+        return self._process_paid_invoices({invoice: status for invoice, status in qr_statuses.items() if status['paid']})
+
+    def _get_qris_qr_statuses(self):
+        """
+        Query the API in order to get updated information on the status of each QR codes linked to the invoices in self.
+
+        :return: a list with the format:
+            {
+                invoice: {
+                    'paid': True,
+                    'qr_statuses': {}
+                },
+                invoice: {
+                    'paid': False
+                }
+            }
+        """
+        result = {}
+        for invoice in self:
+            paid = False
+            data = []
+            # Looping to make requests is far from ideal, but we have no choices as they don't allow getting multiple QR result at once.
+            # Ensure to loop in reverse and check from the most recent QR code.
+            for qr_invoice in reversed(invoice.l10n_id_qris_invoice_details):
+                status_response = invoice.partner_bank_id._qris_fetch_status(qr_invoice)
+                data.append(status_response['data'])
+                if status_response['data'].get('qris_status') == 'paid':
+                    paid = True
+                    break  # We don't really need any unpaid data, as long as we know which QR is paid
```

Comment:
Do we really need to keep the data about non-paid QR?

Reply:
Yes, unless the payment has expired. We don't want to risk discarded data about QR codes that could still be paid.